Add tests for `retryOnEmpty` middleware

Gudahtt opened the pull request:
Tests have been added for the `retryOnEmpty` middleware.

This middleware is only used on the Infura network client, so the tests that demonstrate this retry behavior are only enabled for the `infura` provider type.

Most of the tests added were to cover cases where the retry middleware is skipped, so they were applicable for both provider types.

Closes #17004

## Manual Testing Steps

No functional changes.

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

Gudahtt committed:
Add tests for `retryOnEmpty` middleware

Tests have been added for the `retryOnEmpty` middleware.

This middleware is only used on the Infura network client, so the tests
that demonstrate this retry behavior are only enabled for the `infura`
provider type.

Most of the tests added were to cover cases where the retry middleware
is skipped, so they were applicable for both provider types.

Closes #17004

Gudahtt commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -2859,45 +2938,6 @@ export function testsForRpcMethodSupportingBlockParam(
       });
     });
 
-    it.each([null, undefined, '\u003cnil\u003e'])(
```
This was migrated below to `earliest` block, and to the `block number -> greater than/equal/less than` blocks, because it depended on those factors.

BelfordZ replied:
this will make it more difficult to reflect these changes onto https://github.com/MetaMask/core/pull/1087.

Please limit the diff on these tests to changes that are absolutely necessary.

Gudahtt replied:
Do you have any suggestions on alternative approaches? This was my attempt at making minimal changes while still covering the behavior of this middleware.

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -3204,56 +3308,464 @@ export function testsForRpcMethodSupportingBlockParam(
+          await withMockedCommunications({ providerType }, async (comms) => {
+            const request = {
+              method,
+              params: buildMockParams({ blockParamIndex, blockParam }),
```
Now I kind of wish I hadn't extracted this `blockParam`. It's a bit hard to see just by looking at this `describe` that it's _this_ value we care about being less than the current block number. Would it be clearer if we specified what `blockParam` was in the `less than the current block number`, `equal to the current block number`, and `greater than the current block number` tests? For instance:

``` suggestion
              params: buildMockParams({ blockParamIndex, blockParam: '0x100' }),
```

Gudahtt replied:
Yeah I agree, this is hard to read. It feels wrong to ignore the loop variable as well though.

Maybe it will suffice for now to add a comment pointing out why this is relevant? And we can restructure this later to make more sense.

mcmire replied:
Yeah, if it's a big deal to change this I'm fine with that solution.

Gudahtt replied:
Well, not a big deal, but I just don't want the loop variable to become unused and potentially out-of-sync with what the test is doing. That would also be confusing. And a larger refactor would be beyond what I'd like to do in this PR.

That said, I did find a way to combine at least the first two of these blocks (the two longer ones), hopefully making this more readable as well. And I've added the comments. Done here: 38004d3655ea70bef8a52f3a0e4b3abfc369491a

mcmire reviewed (commented):
Nicely done. Just one had suggestion.

Gudahtt committed:
Improve readability of block number tests

The test cases for passing a block number parameter have been made more
readable. Specifically, a comment has been added each time params are
built to call attention to the block parameter and what value it has,
so that it's clear whether it's larger or smaller than the current
block number.

Additionally the blocks for "less than the current block number" and
"equal to the current block number" have been combined using
`describe.each`.

mcmire commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -3204,56 +3308,311 @@ export function testsForRpcMethodSupportingBlockParam(
-          };
+      describe.each(
+        [
+          ['less than the curent block number', '0x200'],
```
Nit: curent -> current

Gudahtt replied:
Thanks!
```suggestion
          ['less than the current block number', '0x200'],
```

Gudahtt committed:
Fix typo

mcmire reviewed (approved):
Looks good!

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -572,9 +572,31 @@ export function testsForRpcMethodAssumingNoBlockParam(
-  it.each([null, undefined, '\u003cnil\u003e'])(
-    'does not reuse the result of a previous request if it was `%s`',
-    async (emptyValue) => {
+  for (const emptyValue of [null, undefined, '\u003cnil\u003e']) {
```
this will make it more difficult to reflect these changes onto https://github.com/MetaMask/core/pull/1087.

Please limit the diff on these tests to changes that are absolutely necessary.

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -1800,9 +1844,44 @@ export function testsForRpcMethodSupportingBlockParam(
-    it.each([null, undefined, '\u003cnil\u003e'])(
-      'does not reuse the result of a previous request if it was `%s`',
-      async (emptyValue) => {
+    for (const emptyValue of [null, undefined, '\u003cnil\u003e']) {
```
this will make it more difficult to reflect these changes onto https://github.com/MetaMask/core/pull/1087.

Please limit the diff on these tests to changes that are absolutely necessary.

BelfordZ commented on app/scripts/controllers/network/provider-api-tests/shared-tests.js:
```diff
@@ -3204,56 +3308,311 @@ export function testsForRpcMethodSupportingBlockParam(
+      describe.each(
+        [
+          ['less than the current block number', '0x200'],
+          ['equal to the curent block number', '0x100'],
```
is this related to retryOnEmpty or is this a new test that is unrelated?

Gudahtt replied:
Yes, all of these changes are testing functionality in `retryOnEmpty`. The retries covered in this section are the ones referred to in the name "retryOnEmpty".
